src/pam/main.cc

```diff
@@ -0,0 +1,394 @@
+        return PAM_AUTHINFO_UNAVAIL;
+      }
+    }
+
+    globfree(&glob_result);
+  }
+
+  // If enabled, send a notice to the user that facial login is being attempted
+  if (reader.GetBoolean("core", "detection_notice", false)) {
+    if ((pam_res = conv_function(
+             PAM_TEXT_INFO,
+             dgettext("pam", "Attempting facial authentication"))) !=
+        PAM_SUCCESS) {
+      syslog(LOG_ERR, "Failed to send detection notice");
+    }
+  }
+
+  // Get the username from PAM, needed to match correct face model
+  char *user_ptr = nullptr;
+  if ((pam_res = pam_get_user(pamh, (const char **)&user_ptr, nullptr)) !=
+      PAM_SUCCESS) {
+    syslog(LOG_ERR, "Failed to get username");
+    return pam_res;
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+  // We close standard descriptors for the child
+  posix_spawn_file_actions_addclose(&file_actions, STDOUT_FILENO);
+  posix_spawn_file_actions_addclose(&file_actions, STDERR_FILENO);
+  posix_spawn_file_actions_addclose(&file_actions, STDIN_FILENO);
+
+  const char *const args[] = {
+      "/usr/bin/python3", "/lib/security/howdy/compare.py", user_ptr, nullptr};
+  pid_t child_pid;
+
+  // Start the python subprocess
+  if (posix_spawnp(&child_pid, "/usr/bin/python3", &file_actions, nullptr,
+                   (char *const *)args, nullptr) < 0) {
+    syslog(LOG_ERR, "Can't spawn the howdy process: %s", strerror(errno));
+    return PAM_SYSTEM_ERR;
+  }
+
+  mutex m;
+  condition_variable cv;
+  Type confirmation_type;
+  // TODO: Find a clean way to do this
+  Type final_type;
+
+  // This task wait for the status of the python subprocess (we don't want a
+  // zombie process)
+  optional_task<int> child_task(packaged_task<int()>([&] {
+    int status;
+    wait(&status);
+    {
+      unique_lock<mutex> lk(m);
+      confirmation_type = Type::Howdy;
+    }
+    cv.notify_all();
+    return status;
+  }));
+  child_task.activate();
+
+  // This task waits for the password input (if the workaround wants it)
+  optional_task<tuple<int, char *>> pass_task(
+      packaged_task<tuple<int, char *>()>([&] {
+        char *auth_tok_ptr = nullptr;
+        int pam_res = pam_get_authtok(pamh, PAM_AUTHTOK,
+                                      (const char **)&auth_tok_ptr, nullptr);
+        {
+          unique_lock<mutex> lk(m);
+          confirmation_type = Type::Pam;
+        }
+        cv.notify_all();
+        return tuple<int, char *>(pam_res, auth_tok_ptr);
+      }));
+
+  if (auth_tok) {
+    pass_task.activate();
+  }
+
+  // Wait for the end either of the child or the password input
+  {
+    unique_lock<mutex> lk(m);
+    cv.wait(lk);
+    final_type = confirmation_type;
+  }
+
+  if (final_type == Type::Howdy) {
+    // We need to be sure that we're not going to block forever if the
+    // child has a problem
+    if (child_task.wait(3s) == future_status::timeout) {
```

Comment:
Will we ever have to wait so long here? `final_type == Howdy` will only be satisfied when the python process has exited, and all that remains is for the task to return the status. Considering that `future::get` is a blocking operation, you may simply want to call that instead of trying to stop the task.

Reply:
You are definitely right, it's pointless to wait for or kill the child here. It must be an error that I made because of copy-pasting.